### PR TITLE
Migrate NaiveTake and ByteStringBatcher to GraphStage #20288.

### PR DESCRIPTION
**Pull request by danxmoran**
Ref #20288.

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala [hunk 1]
```diff
@@ -47,29 +48,21 @@ class IteratorInterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
       a[NoSuchElementException] should be thrownBy { itr.next() }
     }
 
-    "throw exceptions when chain fails" in {
```
Comment (danxmoran) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala]: The body of this test was exactly the same as the test immediately below.

**Comment by johanandren** [comment]
OK TO TEST

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala [hunk 2]
```diff
@@ -47,29 +48,21 @@ class IteratorInterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
...
+          new GraphStageLogic(shape) with InHandler with OutHandler {
+            override def onPush(): Unit = {
+              val elem = grab(in)
+              if (elem == 2) throw new ArithmeticException()
+              else push(out, elem)
+            }
+            override def onPull(): Unit = pull(in)
+
+            setHandlers(in, out, this)
+          }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala]: couldn't this just be replaced by a `Flow[Int].map(n => if (n == 2) throw new ArithmeticException() else n)`?

**Comment by johanandren** [comment]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala (hunk 2 above)
Comment (danxmoran) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala]: 'Flow[Int].map` returned the wrong type for the `IteratorInterpreter`, but using `Map` directly worked!  PR is updated with the latest.

**Comment by patriknw** [comment]
LGTM, thanks!

**Pull request merged**